### ISSUE TITLE
mAP COCO 2017 results

### ISSUE THREAD

**Issue by thoang3** [fizyr/keras-retinanet#1142]
Hi everyone, 

I'd like to thank the authors of the repo for your great work!

I want to share my training results on COCO 2017. I use default parameters (resnet50 backbone, imagenet pretrained weight...). The training goes quite well, and the loss converges nicely till epoch 50-60. After that mAP stays still as below even though I've tried to train until epoch 96.

Based on the paper, mAP for this architecture should be around 0.35. I am wondering if anybody has been able to replicate that based on this implementation? 
Thank you!

>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.296
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.461
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.313
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.143
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.326
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.398
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.285
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.470
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.513
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.323
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.561
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.663

**Comment by hgaiser** [comment]
Hi, so a long time ago I was able to reproduce their results and got a 0.350 mAP but I haven't been able to reproduce that ever since. I feel like it has to do with how we train (optimized), not what we train on  (the data) or with (the architecture). Reason for that is if you start training using the weights that get 0.350, you will at the first few epochs drop a bit to 0.34-ish and then climb back up to 0.350-ish. Starting a new training process I haven't reached much more than 0.32.

If you have any suggestions or want to test things to see if certain parts can be improved then that is much appreciated. As mentioned, I'd start with looking at the optimizer.

**Comment by AndreaPi** [comment]
If I may, a "relatively" cheap (relatively, because training on COCO is never cheap) test may be to substitute

https://github.com/fizyr/keras-retinanet/blob/3f72a52cb4430febdd0d5f359dfe815baec64d0c/keras_retinanet/bin/train.py#L125

with 

``` 
optimizer = keras.optimizers.SGD(lr=lr, decay=1e-4, momentum=0.9)
``` 
Some other implementations I know of, which claim to match the original paper mAP, use SGD with momentum, instead than Adam. Convergence will be almost surely slower, but the final minimum should have a higher mAP. Also, and maybe more important, it appears that other implementations use some learning rate scheduler. Using a warmup phase and/or a cosine learning rate might help getting an higher mAP.

**Comment by tung-hoang2** [comment]
Thank you for your responses!
I think testing with  SGD is a great way, and I confirm that I was able to achieve 0.34ish mAP using another implementation, which uses SGD as well.
And I agree that learning rate scheduler seems to be quite important too. 
Let me try these and get back to you!

**Comment by thoang3** [comment]
Thank you for your responses!
I think testing with SGD is a great way, and I confirm that I was able to achieve 0.34ish mAP using another implementation, which uses SGD as well.
And I agree that learning rate scheduler seems to be quite important too.
Let me try these and get back to you!

**Comment by hgaiser** [comment]
@thoang3 any update on your tests?

**Comment by thoang3** [comment]
Hi @hgaiser ,

I've tried with SGD optimizer as suggested by @AndreaPi , but the results were quite discouraging, where both mAP and mAP@50 stay near 0 the whole time (attached below). I'll try one more time later this week.

<img width="487" alt="keras-retinanet-SGD" src="https://user-images.githubusercontent.com/9111479/68165455-5a3eba80-ff14-11e9-9a57-bda15e93124f.png">

Moreover, I also tried training again with Adam, and sadly this time the mAP is lower compared to the first time. I guess I might give it another shot next week.

<img width="572" alt="keras-retinanet-Adam" src="https://user-images.githubusercontent.com/9111479/68165543-9245fd80-ff14-11e9-8e3d-5b7c209ffd80.png">
 




> @thoang3 any update on your tests?

**Comment by AndreaPi** [comment]
Which learning rate schedule did you use? The scheduler is very important with SGD. Also, I think the number of epochs needed to get to "convergence" can easily be an order of magnitude larger than for Adam, but that should depend on how close to optimal is your schedule.

**Comment by hgaiser** [comment]
I've also started a training with SGD. It's on epoch 28 (stepsize = 2000, batch_size=4) and this is the evaluation:

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.207
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.343
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.216
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.100
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.232
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.277
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.237
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.413
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.454
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.258
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.503
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.602
```

I started with a learning rate of 0.0025. In the original paper they say they use a learning rate of 0.01, however in another paper (the FPN paper if I recall) they say that they scale the learning rate based on the batch size. Since they have a batch size of 16, I used a fourth of their learning rate. I'll update here if the training reaches an interesting point.

**Comment by hgaiser** [comment]
It got stuck on the following results:

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.231
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.376
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.241
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.110
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.257
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.306
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.252
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.431
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.474
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.277
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.520
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.624
```

I've restarted training starting from those weights, let's see if it gets further this time.

**Comment by thoang3** [comment]
> Which learning rate schedule did you use? The scheduler is very important with SGD. Also, I think the number of epochs needed to get to "convergence" can easily be an order of magnitude larger than for Adam, but that should depend on how close to optimal is your schedule.

@AndreaPi I haven't used learning rate schedule yet, as I'd like to do a testing round without it to see how critical it is. Let me try again with lr schedule.

@hgaiser I agree with you regarding the learning rate based on the batch size. Let me try with lr=0.0025 as well. You didn't use lr schedule either right?

**Comment by hgaiser** [comment]
> @hgaiser I agree with you regarding the learning rate based on the batch size. Let me try with lr=0.0025 as well. You didn't use lr schedule either right?

I didn't use a lr schedule, indeed.

**Comment by hgaiser** [comment]
New training procedure, starting from the previous, so far reached the following results:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.276
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.434
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.292
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.138
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.299
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.373
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.274
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.464
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.507
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.307
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.552
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.665
```

**Comment by hgaiser** [comment]
Latest results after another set of 100 epochs (though the learning rate decreased so much, it might as well have been 25 epochs).

**Comment by reactivetype** [comment]
@hgaiser  could you please share the optimizer and parameters used in latest run?

**Comment by hgaiser** [comment]
Sure:
```shell
python keras_retinanet/bin/train.py --tensorboard-dir logs/clr --batch-size 4 --gpu 1 --epochs 200 --lr 0.0025 --steps 2000 --weights ~/.keras/models/ResNet-50-model.keras.h5 coco /srv/datasets/COCO/
```
Optimizer:
```python
optimizer = keras.optimizers.SGD(lr=lr, decay=1e-4, momentum=0.9)
```
With `lr` set to 0.0025 via the command line.
